File: backend/models/proxy_host.js
// Objection Docs:
// http://vincit.github.io/objection.js/

const db          = require('../db');
const helpers     = require('../lib/helpers');
const Model       = require('objection').Model;
const User        = require('./user');
const AccessList  = require('./access_list');
const Certificate = require('./certificate');
const now         = require('./now_helper');

Model.knex(db);

const boolFields = [
	'is_deleted',
	'ssl_forced',
	'caching_enabled',
	'block_exploits',
	'allow_websocket_upgrade',
	'http2_support',
	'enabled',
	'hsts_enabled',
	'hsts_subdomains',
];

class ProxyHost extends Model {
	$beforeInsert () {
		this.created_on  = now();
		this.modified_on = now();

		// Default for domain_names
		if (typeof this.domain_names === 'undefined') {
			this.domain_names = [];
		}

		// Default for meta
		if (typeof this.meta === 'undefined') {
			this.meta = {};
		}

		this.domain_names.sort();
	}

	$beforeUpdate () {
		this.modified_on = now();

		// Sort domain_names
		if (typeof this.domain_names !== 'undefined') {
			this.domain_names.sort();
		}
	}

	$parseDatabaseJson(json) {
		json = super.$parseDatabaseJson(json);
		return helpers.convertIntFieldsToBool(json, boolFields);
	}

	$formatDatabaseJson(json) {
		json = helpers.convertBoolFieldsToInt(json, boolFields);
		return super.$formatDatabaseJson(json);
	}

	static get name () {
		return 'ProxyHost';
	}

	static get tableName () {
		return 'proxy_host';
	}

	static get jsonAttributes () {
		return ['domain_names', 'meta', 'locations'];
	}

	static get relationMappings () {
		return {
			owner: {
				relation:   Model.HasOneRelation,
				modelClass: User,
				join:       {
					from: 'proxy_host.owner_user_id',
					to:   'user.id'
				},
				modify: function (qb) {
					qb.where('user.is_deleted', 0);
				}
			},
			access_list: {
				relation:   Model.HasOneRelation,
				modelClass: AccessList,
				join:       {
					from: 'proxy_host.access_list_id',
					to:   'access_list.id'
				},
				modify: function (qb) {
					qb.where('access_list.is_deleted', 0);
				}
			},
			certificate: {
				relation:   Model.HasOneRelation,
				modelClass: Certificate,
				join:       {
					from: 'proxy_host.certificate_id',
					to:   'certificate.id'
				},
				modify: function (qb) {
					qb.where('certificate.is_deleted', 0);
				}
			}
		};
	}
}

module.exports = ProxyHost;

File: backend-go/internal/models/proxy_host.go
package models

// ProxyHost represents a reverse proxy configuration
type ProxyHost struct {
	Base
	OwnerUserID    uint   `gorm:"not null" json:"owner_user_id"`
	IsDeleted      uint   `gorm:"default:0" json:"is_deleted"`
	DomainNames    string `gorm:"type:text;not null" json:"domain_names"` // Stored as JSON string array
	ForwardIP      string `gorm:"not null" json:"forward_ip"`
	ForwardPort    uint   `gorm:"not null" json:"forward_port"`
	AccessListID   uint   `gorm:"default:0;not null" json:"access_list_id"`
	CertificateID  uint   `gorm:"default:0;not null" json:"certificate_id"`
	SSLForced      uint   `gorm:"default:0;not null" json:"ssl_forced"`
	CachingEnabled uint   `gorm:"default:0;not null" json:"caching_enabled"`
	BlockExploits  uint   `gorm:"default:0;not null" json:"block_exploits"`
	AdvancedConfig string `gorm:"type:text;not null;default:''" json:"advanced_config"`
	Meta           string `gorm:"type:text;not null" json:"meta"` // Stored as JSON string
}

func (ProxyHost) TableName() string {
	return "proxy_host"
}
